Catch up on lineage pre-commit and GHA updates

dav3r opened the pull request:
<!-- GitHub renders PRs such that soft line breaks are treated as hard
line breaks.  In order to make this template render as expected we
therefore have to avoid soft breaks and therefore will offend
markdownlint with long lines.  This is the reason for the markdownline
disable directive just below.

For more details see:
https://github.com/github/markup/issues/1050#issuecomment-294654762 -->
<!-- markdownlint-disable MD013 -->
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR cherry-picks a lot of commits from https://github.com/cisagov/skeleton-aws-lambda-python/pull/34 in order to get this repo to a healthy place where all of the pre-commit hooks can be run successfully.

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

I plan to update this repo further (move from Python 3.9 to 3.13 and use a current version of NodeJS) in a future PR, but I needed to get current on everything in this PR first.

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

All pre-commit hooks are passing.

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] All new and existing tests pass.

mcdonnnj committed:
Adjust the indentation rule for yamllint

Use a specific number of spaces instead of the default of only caring
if the number of spaces used is consistent within a file. Ensure that
block sequences inside of mappings are indented.

mcdonnnj committed:
Update yamllint to disallow non-empty flow collection styles

The use of flow sequences and mappings is not as readable as block
collections and so should be discouraged. Since it is a cleaner
representation for empty collections we will allow those, but if an
application otherwise requires flow collections they can be explicitly
enabled by disabling the checks per
https://yamllint.readthedocs.io/en/stable/disable_with_comments.html

mcdonnnj committed:
Add a pre-commit hook to run pip-audit

The pip-audit tool will audit any supplied pip requirements files for
vulnerable packages.

Commit:
Add checks for semantic python versions

mcdonnnj committed:
Add a meta hook to the pre-commit configuration

Add the `check-useless-excludes` meta hook to verify that any defined
`exclude` directives apply to at least one file in the repository.

mcdonnnj committed:
Remove `exclude` directive that does not apply to any files

jsf9k committed:
Pin ansible-core when running the ansible-lint linter

New versions of ansible-core (2.16.7 and 2.17.0) have been released
that do not suffer from the bug discussed in ansible/ansible#82702.
This bug broke any symlinked files in vars, tasks, etc. for any
Ansible role installed via ansible-galaxy.

All versions later than ansible-core 2.16.7 and 2.17.0 should function
as expected.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Remove unnecessary line

The line is not only unnecessary, it was commented out to boot!

jsf9k committed:
Explain why ansible may need to be added as a dependency for ansible-lint

On its own ansible-lint does not pull in ansible, only ansible-core.
Therefore, if an Ansible module lives in ansible instead of
ansible-core, the linter will complain that the module is unknown.  In
these cases it is necessary to add the ansible package itself as an
additional dependency, with the same pinning as is done in
requirements-test.txt of cisagov/skeleton-ansible-role.

mcdonnnj committed:
Update pre-commit hook versions

This is done automatically with the `pre-commit autoupdate` command.
The pre-commit/mirrors-prettier was manually held back because the
latest tags are for alpha releases of the next major version.

mcdonnnj committed:
Manually update the prettier hook

Use the latest v3 release available from NPM.

jsf9k committed:
Correct commented-out ansible pin

The pin now agrees with what is in cisagov/skeleton-ansible-role.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Add a lower-bound pin for flake8-docstrings

mcdonnnj committed:
Update pre-commit hook versions

This is done automatically with the `pre-commit autoupdate` command.
The pre-commit/mirrors-prettier hook was manually held back because the
latest tags are for alpha releases of the next major version.

mcdonnnj committed:
Manually update the prettier hook

Use the latest v3 release available from NPM.

mcdonnnj committed:
Use the rbubley/mirrors-prettier hook for prettier

This replaces the now archived pre-commit/mirrors-prettier hook.

mcdonnnj committed:
Add a new trigger for the sync-labels GitHub Actions workflow

Add a `workflow_dispatch` trigger so we can manually run the workflow
if needed.

mcdonnnj committed:
Remove unnecessary quotes in the sync-labels workflow

jsf9k committed:
Remove @jasonodoom as a codeowner

He is no longer a member of @cisagov/vm-dev.

jsf9k committed:
Pin to a specific version

Previously we only provided a lower bound for the version, but pinning to a specific version aligns with what has been done with the prettier hook and how pre-commit hooks are pinned in general.

The flake8-docstrings package is rarely updated, so there is no real downside to pinning to a specific version.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj committed:
Add four new hooks from pre-commit/pre-commit-hooks

dependabot[bot] committed:
Bump actions/cache from 3 to 4

Bumps [actions/cache](https://github.com/actions/cache) from 3 to 4.
- [Release notes](https://github.com/actions/cache/releases)
- [Changelog](https://github.com/actions/cache/blob/main/RELEASES.md)
- [Commits](https://github.com/actions/cache/compare/v3...v4)

---
updated-dependencies:
- dependency-name: actions/cache
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump crazy-max/ghaction-github-status from 3 to 4

Bumps [crazy-max/ghaction-github-status](https://github.com/crazy-max/ghaction-github-status) from 3 to 4.
- [Release notes](https://github.com/crazy-max/ghaction-github-status/releases)
- [Commits](https://github.com/crazy-max/ghaction-github-status/compare/v3...v4)

---
updated-dependencies:
- dependency-name: crazy-max/ghaction-github-status
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

mcdonnnj committed:
Update pre-commit hook versions

This is done automatically with the `pre-commit autoupdate` command.

mcdonnnj committed:
Restrict permissions of GITHUB_TOKEN

This changes the default permissions for the GITHUB_TOKEN used in our
GitHub Actions configuration to the minimum required to successfully
run.

mcdonnnj committed:
Sort hook ids in each pre-commit hook entry

Ensure that all hook ids are sorted alphabetically in each hook entry
in our pre-commit configuration.

jsf9k committed:
Bump up the lower bound on ansible-core

This is being done because the pip-audit pre-commit hook identifies a
vulnerability in ansible-core version 2.16.13.  Note that this
requires that we bump up ansible to version 10 since all versions of
ansible 9 have a dependency on ~=2.16.X.

jsf9k committed:
Update the version of the ansible-lint pre-commit hook

Version 24.10.0 is the first version that supports Fedora 41 as a
valid platform.

jsf9k committed:
Adjust pin for ansible-core

The pin of ansible-core was originally put in place because the
pip-audit pre-commit hook identifies a vulnerability in ansible-core
2.16.13.  Normally we would pin ansible-core to >2.16.13, but in the
spirit of the earlier, optional pin of ansible>=10 we pin ansible-core
to >=2.17.  This effectively also pins ansible to >=10.

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Add comments about looming EOL issues for ansible and ansible-core

This adds even more evidence for why it is a good idea to go ahead and
upgrade ansible and ansible-core, in addition to the vulnerability
that pip-audit turned up.

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj committed:
Add yamllint configuration settings to appease ansible-lint

When running ansible-lint it will throw the following warning with our
current configuration:
WARNING  Found incompatible custom yamllint configuration (.yamllint), please either remove the file or edit it to comply with:
  - comments.min-spaces-from-content must be 1
  - braces.max-spaces-inside must be 1
  - octal-values.forbid-implicit-octal must be true
  - octal-values.forbid-explicit-octal must be true.
Thus we implement these configuration rules.

mcdonnnj committed:
Re-enable the yamllint truthy rule

Previously we disabled the `truthy` rule due to Ansible's use of
`yes`/`no` for boolean values. That is no longer the case and the
default configuration used by ansible-lint now has this rule enabled.
The use of `on` as a key in GitHub Actions workflow syntax means we
needed to add disable-line comments for the truthy rule.

jsf9k committed:
Apply our standard job preamble via cisagov/action-job-preamble

This new action simply applies our standard permissions monitoring and
runner hardening.  Using it allows us to DRY out the GH Actions
workflows in our skeleton repositories a bit.

jsf9k committed:
Set actions_permissions_config input

jsf9k committed:
Add a friendly name to the cisagov/action-job-preamble steps

mcdonnnj committed:
Update pre-commit hook versions

This is done automatically with the `pre-commit autoupdate` command.

jsf9k committed:
Update input name for cisagov/action-job-preamble

jsf9k committed:
Use cisagov/action-job-preamble in sync-labels.yml workflow

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Use v1 tag of cisagov/action-job-preamble

jsf9k committed:
Re-add comment explaining where the org var comes from

Flesh out the comment a little so its meaning is clearer.

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Flesh out org var comment even more

Make sure to mention that the permissions monitoring config can be
changed by creating a repo-level variable; there is no need to modify
the workflow.

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Add a workflow to run actions/dependency-review-action

This action reviews dependency changes for vulnerabilities and license
changes.

jsf9k committed:
Update comment to match what is in cisagov/skeleton-docker

Also correctly sort YAML keys.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Use cisagov/action-job-preamble

This aligns with the changes in cisagov/skeleton-generic#201.

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Use cisagov/action-job-preamble

This aligns with the changes in cisagov/skeleton-generic#201.

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Use cisagov/action-job-preamble instead of separate actions

Use cisagov/action-job-preamble instead of
crazy-max/ghaction-github-status and crazy-max/ghaction-dump-context
directly.

jsf9k committed:
Use cisagov/action-job-preamble instead of separate actions

Use cisagov/action-job-preamble instead of
crazy-max/ghaction-github-status and crazy-max/ghaction-dump-context
directly.

Also disable permissions monitoring since it is poorly implemented and
has been causing a lot of problems due to hogging or leaking memory.

jsf9k committed:
Do not disable GitHub permissions monitoring by default

But do leave a commented-out line that can be uncommented to do so.
The idea is that we should only comment out this functionality where
we really must.

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Disable GH permissions monitoring everywhere

This
functionality (https://github.com/GitHubSecurityLab/actions-permissions/tree/main/monitor)
is poorly implemented and has been causing problems due to the MITM
implementation hogging or leaking memory.

This functionality should be re-enabled when practical.  See
cisagov/skeleton-generic#207 for more details.

jsf9k committed:
Fix typo that was copied and pasted all over the show

Co-authored-by: David Harris <123905168+dv4harr10@users.noreply.github.com>

jsf9k committed:
Fix typo

jsf9k committed:
Upgrade to the latest version of the ansible-lint pre-commit hook

Version 25.4.0 is the first version to support Fedora 42 in the
Ansible YAML metadata schema.

mcdonnnj committed:
Update the version of the second usage of bandit

We use the bandit pre-commit hook twice in this configuration and only
one of the configurations is updated automatically from upstream.
Therefore, we must manually update the second configuration to keep the
versions in sync.

mcdonnnj committed:
Remove unnecessary shebang

mcdonnnj committed:
Consistently use cisagov/action-job-preamble

Update the `test` and `build` jobs to also use the
cisagov/action-job-preamble action.

mcdonnnj committed:
Adjust artifact name in `build` workflow

This is necessary to have a valid artifact name for upload.

dav3r committed:
Use the hashicorp/setup-packer GitHub Action

Instead of manually installing Packer we can instead leverage the
hashicorp/setup-packer Action just as we do for Terraform.

dav3r committed:
Add the GitHubSecurityLab/actions-permissions/monitor Action

This Action will provide information about the usage of GITHUB_TOKEN in
the workflow. It should be added to _every_ job in _any_ workflow to
provide information for analysis.

dav3r committed:
Update the commented out dependabot ignore directives

Add a directive for hashicorp/setup-packer that was missed when it was
added to the `build` workflow. Add a directive for
cisagov/setup-env-github-action that is not strictly necessary since we
currently just pull from the `develop` branch, but is good to have in
case we were to change that in the future.

dav3r committed:
Configure quoted strings rule for yamllint

Add a configuration for the `quoted-strings` rule that matches our best
practices. Other files are updated to comply with these new settings.

dav3r committed:
Add bump_version script

Also add semver as a dev requirement.

I'd like to start versioning descendants of skeleton-ansible-role (in
anticipation of pinning Ansible role versions at a future date), and I
thought it would make sense to go ahead and implement this at the
skeleton-generic level to force us to start versioning all
repositories.

Repositories that already version can ignore these changes when they
flow down via Lineage, since they will already have their own version
files and version-bumping script.

dav3r committed:
Add a Dependabot ignore directive for cisagov/action-job-preamble

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Remove two Dependabot ignore directives

GitHubSecurityLab/actions-permissions and step-security/harden-runner
are no longer direct dependencies since we are now using
cisagov/action-job-preamble.

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Subsume GH status checks and context dumping into cisagov/action-job-preamble

This action supports this functionality now, so we may as well take
advantage of it.

Also disable GH permissions monitoring, since that functionality is
poorly implemented and has been causing a lot of problems due to the
MITM implementation hogging or leaking memory.

dav3r committed:
Add a CodeQL workflow to this repository

CodeQL now supports GitHub Actions as a language, so it makes sense to
add such a workflow to this repository.

See this link for more details:
https://github.blog/changelog/2024-12-17-find-and-fix-actions-workflows-vulnerabilities-with-codeql-public-preview/

dav3r committed:
Add the CodeQL action to the Dependabot configuration

Children of this skeleton repository will require this Dependabot
ignore directive.

dav3r committed:
Add the dependency review action to the Dependabot configuration

Children of this skeleton repository will require this Dependabot
ignore directive.

dav3r committed:
Update files to comply with new yamllint rules

dav3r committed:
Update the version management configuration

Remove the `bump_version.sh` script as it has been superceded by the
`bump-version` script, remove the version.txt inherited from upstream,
update the `bump-version` script to modify the correct version-tracking
file, and update the src/version.txt file to match the updated format
for version-tracking files that are not imported into code.

dav3r committed:
Appease yamllint

jsf9k commented on docker-compose.yml:
```diff
@@ -1,5 +1,5 @@
 ---
-version: '3.2'
+version: "3.2"
```
The `version` tag is actually deprecated if you want to go ahead and remove it.

dav3r replied:
Thanks, removed in https://github.com/cisagov/publish-egress-ip-lambda/pull/49/commits/9163c3117227d54660dd801ada848fd36c75b0f4.

jsf9k reviewed (approved):
Approved, with one comment that you can take or leave as you see fit.

dav3r committed:
Remove deprecated "version" tag

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

dav3r committed:
Grab latest version of setup-env from cisagov/skeleton-generic

I missed some commits while cherry-picking for this PR.  Rather than try to sort that out, it's simpler to just grab the whole file and include the changes here.

dav3r committed:
Remove duplicated dependabot ignore directive
